File: src/ragkor/match.ts
/** RAGKOR 매처 — 사전 + 자모 유사도 + BM25를 하나로 묶는다. */
import { similarity } from "./jamo.js";
import { buildLexicon, type Cluster, type LexiconData } from "./lexicon.js";
import { normalizeToken, tokenize } from "./normalize.js";

/** 표면형 → 대표형 사전. 규칙으로 생성하므로 외부 파일이 필요 없다. */
export class Lexicon {
  readonly canon: Record<string, string>;
  readonly origin: Record<string, string>;
  readonly clusters: Cluster[];
  readonly entries: number;
  private readonly members = new Map<string, Set<string>>();
  private readonly byLen = new Map<number, string[]>();

  constructor(data: LexiconData = buildLexicon()) {
    this.canon = data.canon;
    this.origin = data.origin;
    this.clusters = data.clusters;
    this.entries = Object.keys(this.canon).length;

    for (const c of this.clusters) {
      for (const m of c.members) this.addMember(c.canon, m);
    }
    for (const [surface, canonical] of Object.entries(this.canon)) {
      this.addMember(canonical, surface);
      // 자모 유사도 폴백용 — 사전에 없는 오타를 잡을 때 후보를 좁힌다
      const bucket = this.byLen.get(surface.length);
      if (bucket) bucket.push(surface);
      else this.byLen.set(surface.length, [surface]);
    }
  }

  private addMember(canonical: string, surface: string) {
    const set = this.members.get(canonical);
    if (set) set.add(surface);
    else this.members.set(canonical, new Set([surface]));
  }

  /** 표면형을 대표형으로. 사전에 없으면 정규화형을 그대로 돌려준다. */
  canonical(token: string): string {
    const t = normalizeToken(token);
    const hit = this.canon[t];
    if (hit) return hit;
    return this.canon[token.trim().toLowerCase()] ?? t;
  }

  /** 사전에 없는 오타를 자모 거리로 가장 가까운 표제어에 붙인다. */
  fuzzyCanonical(token: string, threshold = 0.86): { canon: string; score: number } {
    const t = normalizeToken(token);
    if (this.canon[t]) return { canon: this.canon[t]!, score: 1 };
    let best = t;
    let score = 0;
    for (const L of [t.length - 1, t.length, t.length + 1]) {
      for (const cand of this.byLen.get(L) ?? []) {
        const s = similarity(t, cand);
        if (s > score) {
          best = cand;
          score = s;
          if (s >= 0.99) break;
        }
      }
    }
    return score >= threshold ? { canon: this.canon[best] ?? best, score } : { canon: t, score: 0 };
  }

  /** 검색 확장용 — 같은 뜻으로 쓰이는 표면형을 전부 돌려준다. */
  expand(token: string): Set<string> {
    const c = this.canonical(token);
    const out = new Set<string>([c, normalizeToken(token)]);
    for (const m of this.members.get(c) ?? []) out.add(m);
    out.delete("");
    return out;
  }

  /** 텍스트 → 대표형 토큰 열. 비교·검색은 전부 이 형태에서 한다. */
  analyze(text: string): string[] {
    return tokenize(text).map((t) => this.canonical(t));
  }
}

let cached: Lexicon | null = null;

/** 프로세스 하나당 한 번만 만든다 (생성 ~100ms). */
export function defaultLexicon(): Lexicon {
  if (!cached) cached = new Lexicon();
  return cached;
}

/** 청크 검색용. 유의어 확장을 질의 쪽에 적용해 재현율을 올린다. */
export class BM25 {
  private readonly docs: Map<string, number>[];
  private readonly lens: number[];
  private readonly avg: number;
  private readonly df = new Map<string, number>();
  private readonly N: number;

  constructor(
    readonly chunks: string[],
    private readonly lex: Lexicon = defaultLexicon(),
    private readonly k1 = 1.4,
    private readonly b = 0.72,
  ) {
    this.docs = chunks.map((c) => {
      const m = new Map<string, number>();
      for (const t of this.lex.analyze(c)) m.set(t, (m.get(t) ?? 0) + 1);
      return m;
    });
    this.lens = this.docs.map((d) => [...d.values()].reduce((a, x) => a + x, 0) || 1);
    this.avg = this.lens.reduce((a, x) => a + x, 0) / Math.max(1, this.lens.length);
    for (const d of this.docs) {
      for (const k of d.keys()) this.df.set(k, (this.df.get(k) ?? 0) + 1);
    }
    this.N = Math.max(1, this.docs.length);
  }

  private idf(term: string): number {
    const n = this.df.get(term) ?? 0;
    return Math.log(1 + (this.N - n + 0.5) / (n + 0.5));
  }

  search(query: string, top = 5): { index: number; score: number }[] {
    const terms = new Set<string>();
    for (const t of tokenize(query)) {
      for (const x of this.lex.expand(t)) terms.add(this.lex.canonical(x));
    }
    const scores: { index: number; score: number }[] = [];
    this.docs.forEach((d, i) => {
      let s = 0;
      for (const term of terms) {
        const f = d.get(term) ?? 0;
        if (!f) continue;
        s += this.idf(term) * (f * (this.k1 + 1)) /
          (f + this.k1 * (1 - this.b + this.b * this.lens[i]! / this.avg));
      }
      if (s > 0) scores.push({ index: i, score: s });
    });
    scores.sort((a, b) => b.score - a.score);
    return scores.slice(0, top);
  }
}

File: ragkor/match.py
"""RAGKOR 매처 — 사전 + 자모 유사도 + BM25를 하나로 묶는다."""
import json
import math
import re
from collections import Counter, defaultdict
from pathlib import Path
from typing import Dict, Iterable, List, Optional, Set, Tuple

from .jamo import similarity
from .normalize import fold, ngrams, normalize_token, numbers_in, sentences, tokenize

_LEX_PATH = Path(__file__).parent / "lexicon.json"


class Lexicon:
    """표면형 → 대표형 사전. 없으면 규칙으로 즉석 생성한다(Colab 단일파일 대비)."""

    def __init__(self, data: Optional[dict] = None):
        if data is None:
            data = self._load()
        self.canon: Dict[str, str] = data.get("canon", {})
        self.origin: Dict[str, str] = data.get("origin", {})
        self.clusters: List[dict] = data.get("clusters", [])
        self.entries = len(self.canon)
        self._members: Dict[str, Set[str]] = defaultdict(set)
        for c in self.clusters:
            for m in c["members"]:
                self._members[c["canon"]].add(m)
        for surface, canonical in self.canon.items():
            self._members[canonical].add(surface)
        # 자모 유사도 폴백용 — 사전에 없는 오타를 잡을 때 후보를 좁힌다
        self._by_len: Dict[int, List[str]] = defaultdict(list)
        for s in self.canon:
            self._by_len[len(s)].append(s)

    @staticmethod
    def _load() -> dict:
        if _LEX_PATH.exists():
            return json.loads(_LEX_PATH.read_text(encoding="utf-8"))
        from .build_lexicon import build
        return build()

    def canonical(self, token: str) -> str:
        """표면형을 대표형으로. 사전에 없으면 정규화형을 그대로 돌려준다."""
        t = normalize_token(token)
        if t in self.canon:
            return self.canon[t]
        raw = token.strip().lower()
        return self.canon.get(raw, t)

    def fuzzy_canonical(self, token: str, threshold: float = 0.86) -> Tuple[str, float]:
        """사전에 없는 오타를 자모 거리로 가장 가까운 표제어에 붙인다."""
        t = normalize_token(token)
        if t in self.canon:
            return self.canon[t], 1.0
        best, score = t, 0.0
        for L in (len(t) - 1, len(t), len(t) + 1):
            for cand in self._by_len.get(L, ()):
                s = similarity(t, cand)
                if s > score:
                    best, score = cand, s
                    if s >= 0.99:
                        break
        if score >= threshold:
            return self.canon.get(best, best), score
        return t, 0.0

    def expand(self, token: str) -> Set[str]:
        """검색 확장용 — 같은 뜻으로 쓰이는 표면형을 전부 돌려준다."""
        c = self.canonical(token)
        out = {c, normalize_token(token)}
        out |= self._members.get(c, set())
        return {x for x in out if x}

    def analyze(self, text: str) -> List[str]:
        """텍스트 → 대표형 토큰 열. 비교·검색은 전부 이 형태에서 한다."""
        return [self.canonical(t) for t in tokenize(text)]


_DEFAULT: Optional[Lexicon] = None


def default_lexicon() -> Lexicon:
    global _DEFAULT
    if _DEFAULT is None:
        _DEFAULT = Lexicon()
    return _DEFAULT


class BM25:
    """청크 검색용. 유의어 확장을 질의 쪽에 적용해 재현율을 올린다."""

    def __init__(self, chunks: List[str], lex: Optional[Lexicon] = None,
                 k1: float = 1.4, b: float = 0.72):
        self.lex = lex or default_lexicon()
        self.chunks = chunks
        self.k1, self.b = k1, b
        self.docs = [Counter(self.lex.analyze(c)) for c in chunks]
        self.lens = [sum(d.values()) or 1 for d in self.docs]
        self.avg = sum(self.lens) / max(1, len(self.lens))
        self.df: Counter = Counter()
        for d in self.docs:
            self.df.update(d.keys())
        self.N = max(1, len(self.docs))

    def _idf(self, term: str) -> float:
        n = self.df.get(term, 0)
        return math.log(1 + (self.N - n + 0.5) / (n + 0.5))

    def search(self, query: str, top: int = 5) -> List[Tuple[int, float]]:
        terms: Set[str] = set()
        for t in tokenize(query):
            terms |= {self.lex.canonical(x) for x in self.lex.expand(t)}
        scores = []
        for i, d in enumerate(self.docs):
            s = 0.0
            for term in terms:
                f = d.get(term, 0)
                if not f:
                    continue
                s += self._idf(term) * (f * (self.k1 + 1)) / (
                    f + self.k1 * (1 - self.b + self.b * self.lens[i] / self.avg))
            if s > 0:
                scores.append((i, s))
        scores.sort(key=lambda x: -x[1])
        return scores[:top]
